router: scoped RDS (2/4): support scoped routing configuration

AndresGuedez opened the pull request:
This PR introduces:
- Support for scoped routing configuration parsing.
- Inline and dynamic config providers using the ConfigProvider framework (introduced in #5243).
- Unit and integration testing scaffolding.

This is the second PR in the chain to implement Scoped RDS (see #4704 for design).

NOTE: this PR does not actually implement the scoped routing logic; it will be introduced in the next PR in this chain.

*Risk Level*: Low; this API is not yet fully implemented and is config guarded (defaulting to disabled)
*Testing*: Unit and integration tests added
*Docs Changes*: SRDS (ScopedRouteConfigurationsSet) proto documentation

snowp commented on test/common/router/scoped_rds_test.cc:
```diff
@@ -0,0 +1,296 @@
+  parseScopedRoutesConfigFromYaml(*resources.Add(), config_yaml);
+  EXPECT_THROW(subscription.onConfigUpdate(resources, "1"), ProtoValidationException);
+
+  // 'scope_key_builder.fragments' validation
```
Mind including comments about what is invalid about the configuration? Here and for other invalid config tests. Might also be valuable to verify the exception message to ensure that we're tickling the correct validation

AndresGuedez replied:
Done.

snowp commented on test/common/router/scoped_rds_test.cc:
```diff
@@ -0,0 +1,296 @@
+  ~ScopedRoutesConfigProviderManagerTest() override = default;
+};
+
+// Tests that the /configdump handler returns the corresponding scoped routing config.
```
nit: /config_dump

AndresGuedez replied:
Done.

snowp commented on test/extensions/filters/network/http_connection_manager/config_test.cc:
```diff
@@ -447,10 +458,142 @@ TEST_F(HttpConnectionManagerConfigTest, ReversedEncodeOrderConfig) {
+  - name: envoy.router
+)EOF";
+
+  HttpConnectionManagerConfig(parseHttpConnectionManagerFromV2Yaml(config_yaml), context_,
```
use `EXPECT_NO_THROW` to indicate that we don't expect this to throw

AndresGuedez replied:
Done.

mattklein123 commented on api/envoy/api/v2/srds.proto:
```diff
@@ -0,0 +1,153 @@
+
+        oneof extract_type {
+          // Specifies the index of the element to extract.
+          int32 index = 3;
```
unsigned?

AndresGuedez replied:
I didn't find mention of signed vs unsigned in Envoy's style guide so I defaulted to Google's style of using signed integers and relying on assertions to check val >= 0.

mattklein123 commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -389,11 +398,31 @@ message Rds {
+
+    // Must be set to true when scoped RDS is enabled.
+    // If this is enabled, RDS subscriptions will be triggered by scoped RDS config updates using
+    // this configuration as a template.
```
+1

AndresGuedez replied:
Per https://github.com/envoyproxy/envoy/pull/5839#discussion_r257882809, I am changing this and will be getting rid of the scoped_rds_template. Instead, the ScopedRds message in HttpConnectionManager will include the ConfigSource required for the RDS server along with the ConfigSource for the SRDS server.

mattklein123 commented on api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:
```diff
@@ -63,6 +64,14 @@ message HttpConnectionManager {
 
+  oneof scoped_routes_specifier {
+    // Configuration for the Scoped RDS API which dynamically loads routing configuration scopes.
+    ScopedRds scoped_rds = 30;
```
Would it be simpler if these options were just in the `route_specifier` oneof above? Why have them be orthogonal?

AndresGuedez replied:
My initial intention keeping them separate was to avoid confusion that SRDS and RDS are tightly coupled and require the same config source, etc.

However, based on the feedback in this review and reassessing this from the user's standpoint, I agree this is simpler if `ScopedRds` becomes part of the `route_specifier`, and `ScopedRds` expands to include both the `ConfigSource` for RDS and SRDS. This would allow me to remove `scoped_rds_template` from `Rds`, which avoids the confusion related to that field (I've never been quite happy with that field since I started working on this).

mattklein123 commented on api/envoy/api/v2/srds.proto:
```diff
@@ -0,0 +1,153 @@
+  }
+}
+
+// This configuration represents a set of "scopes", each containing independent routing
```
I'm going to be honest in saying that it's very hard for me to follow the messages/documentation in this file. I read the spec a while ago but I've forgotten it. Would it be possible to add some high level documentation here that walks the end user through some examples of how the API is setup, the fragment builder works, etc. If you plan on doing this in arch overview docs that's fine too, whatever is easiest. The reason I ask for this now is I think it would make it a lot easier to review.

AndresGuedez replied:
My initial plan was to add in-depth documentation including examples to the arch overview docs (which I wasn't planning to do until the feature is fully implemented in PR 3), but I agree better documentation in the proto definition would be helpful for this review. I am working on this.

mattklein123 replied:
Thanks a ton. I think even just some YAML examples here in the proto docs would be super helpful. You could move them to the arch overview later or just link to them here, so hopefully zero wasted effort.

AndresGuedez commented:
@snowp @mattklein123 Thanks for the feedback and patience. I wasn't able to spend time on this last week but I will be actively addressing the comments and updating the PR this week.

AndresGuedez commented on api/envoy/api/v2/srds.proto:
```diff
@@ -0,0 +1,153 @@
+//   The Scoped RDS API is not yet fully implemented and *should not* be enabled in
+//   :ref:`envoy_api_msg_config.filter.network.http_connection_manager.v2.HttpConnectionManager`.
+
+// The resource_names field in DiscoveryRequest specifies a set of route configuration scopes. Each
```
Done.

AndresGuedez commented on api/envoy/api/v2/srds.proto:
```diff
@@ -0,0 +1,153 @@
+  }
+}
+
+// This configuration represents a set of "scopes", each containing independent routing
```
Done.

AndresGuedez commented on api/envoy/api/v2/srds.proto:
```diff
@@ -0,0 +1,153 @@
+      //
+      // .. code::
+      //
+      //    X-Header: a,b;c,d
```
Done.

AndresGuedez commented on api/envoy/api/v2/srds.proto:
```diff
@@ -0,0 +1,153 @@
+    // Specifies the mechanism for constructing fragments which are composed into keys.
+    message FragmentBuilder {
+      // Specifies how the value of a header should be extracted.
+      // The following example maps the structure of a header to the fields in this message.
```
Added an example in the ScopedRouteConfigurationsSet documentation.

AndresGuedez commented on api/envoy/api/v2/srds.proto:
```diff
@@ -0,0 +1,153 @@
+      }
+    }
+
+    // The constructed key consists of the union of these fragments.
```
Yes, updated comment.

AndresGuedez commented on api/envoy/api/v2/srds.proto:
```diff
@@ -0,0 +1,153 @@
+  // :ref:`Key <envoy_api_msg_ScopedRouteConfigurationsSet.Scope.Key>` which is matched against
+  // each HTTP request.
+  message Scope {
+    // Specifies a key which is matched against by the output of a :ref:`ScopeKeyBuilder
```
Done.

AndresGuedez commented on api/envoy/api/v2/srds.proto:
```diff
@@ -0,0 +1,153 @@
+          option (validate.required) = true;
+
+          // A string to match against.
+          string string_key = 1;
```
The comment a few lines above pertaining to the 'Key' message as a whole attempts to clarify this. Let me know if it is sufficient.

AndresGuedez commented on include/envoy/router/scopes.h:
```diff
@@ -0,0 +1,30 @@
+
+  /**
+   * Based on the incoming HTTP request headers, returns the configuration to use for selecting a
+   * target route (to either a route entry or a direct response entry).
```
Done.

AndresGuedez commented on source/common/config/config_provider_impl.h:
```diff
@@ -173,7 +173,10 @@ class ConfigSubscriptionInstanceBase : public Init::Target,
    */
-  void onConfigUpdateFailed() { runInitializeCallbackIfAny(); }
+  void onConfigUpdateFailed() {
+    setLastUpdated();
```
Yes. This is a bug I spotted while working on this. I bundled it with this PR since there are no other users of ConfigSubscriptionInstanceBase yet.

AndresGuedez commented on source/common/http/conn_manager_config.h:
```diff
@@ -252,10 +253,17 @@ class ConnectionManagerConfig {
+
+  /**
+   * @return Config::ConfigProvider* the configuration provider used to acquire scoped routing
+   * configuration for each request flow. Pointer ownership is _not_ transferred to the caller of
```
The `route_config_provider_` returned by this accessor can now be `nullptr`, since enabling scoped RDS means that an HCM scope RDS RouteConfigProvider is no longer necessary.

Another option would be to use `absl::optional<>` but I felt `unique_ptr` is a better fit due to how the delayed allocation happens via `Router::RouteConfigProvider::create()`.

I added the comment about pointer ownership out of an abundance of caution; given Envoy's style of using `unique_ptr`s for ownership transfers, I wouldn't expect a consumer of this interface to confuse the purpose of this API.

AndresGuedez commented on source/common/router/scoped_config_impl.h:
```diff
@@ -0,0 +1,47 @@
+  ScopedConfigImpl(const envoy::api::v2::ScopedRouteConfigurationsSet& config_proto)
+      : name_(config_proto.name()) {}
+
+  ~ScopedConfigImpl() override = default;
```
Removed.

AndresGuedez commented on source/common/router/scoped_rds.cc:
```diff
@@ -0,0 +1,191 @@
+  const auto& scoped_routes_config = resources[0];
+  MessageUtil::validate(scoped_routes_config);
+  // The name we receive in the config must match the name requested.
+  if (!(scoped_routes_config.name() == scoped_routes_config_name_)) {
```
Done.

AndresGuedez commented on source/extensions/filters/network/http_connection_manager/config.cc:
```diff
@@ -64,11 +65,48 @@ std::unique_ptr<Http::InternalAddressConfig> createInternalAddressConfig(
+
+    if (config.rds().subscription_specifier_case() !=
+            envoy::config::filter::network::http_connection_manager::v2::Rds::kScopedRdsTemplate ||
+        config.rds().scoped_rds_template() != true) {
```
Per my replies to a few other comments, the `scoped_rds_template` is going away.

AndresGuedez commented on source/extensions/filters/network/http_connection_manager/config.cc:
```diff
@@ -153,9 +200,19 @@ HttpConnectionManagerConfig::HttpConnectionManagerConfig(
                                                                          context_.listenerScope())),
       proxy_100_continue_(config.proxy_100_continue()),
       delayed_close_timeout_(PROTOBUF_GET_MS_OR_DEFAULT(config, delayed_close_timeout, 1000)) {
+  // Throws an exception on failure.
```
Done.

AndresGuedez commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -74,8 +74,25 @@ class HttpConnectionManagerImplTest : public TestBase, public ConnectionManagerC
     std::shared_ptr<Router::MockConfig> route_config_{new NiceMock<Router::MockConfig>()};
   };
 
+  struct ScopedRouteConfigProvider : public Config::ConfigProvider {
```
Yes. I've now pulled this out into a common implementation included by both this test and the fuzz_test.

AndresGuedez commented:
Apologies for the delay responding to feedback.

I have addressed most of the review comments. I am now working on removing the `scoped_rds_template` per https://github.com/envoyproxy/envoy/pull/5839#discussion_r257882809. This should simplify the logic in HttpConnectionManager and also greatly clarify the interaction between SRDS and RDS ConfigSource configuration.

AndresGuedez commented on api/envoy/api/v2/srds.proto:
```diff
@@ -0,0 +1,214 @@
+// which Envoy would fetch via RDS.
+//
+// [#comment:next free field: 4]
+message ScopedRouteConfigurationsSet {
```
Upon revisiting the SRDS proto in the context of Incremental xDS, I have decided to make a few changes after discussing with @htuch.

To easily enable incremental updates to the routing scopes (represented by the `Scope` message) by leveraging the existing Incremental xDS API, I am going to move the `ScopeKeyBuilder` specification out of the `ScopedRouteConfigurationsSet` into the `HttpConnectionManager`, and instead of publishing a `repeated Scope scopes = ` in this message, each SRDS proto (which will most likely be renamed to `ScopedRouteConfiguration`) will contain a single `Scope`.

This change creates much better alignment with the API/protocol already defined for incremental updates, and also simplifies the SRDS API.

I am actively working on this now and will introduce the changes discussed as part of the next few commits to this PR.

mattklein123 replied:
@AndresGuedez cool I will defer looking at the API again until this is done. Thank you for iterating on this.

/wait

AndresGuedez commented:
I have made the changes to modify the SRDS API to be delta friendly; what is the best way to move forward with this review?

Can this PR be reopened?

htuch commented:
@AndresGuedez can you merge master to resolve conflicts?

AndresGuedez commented:
> @AndresGuedez can you merge master to resolve conflicts?

Merged master and pushed commit with cleanup of the ConfigProvider framework.

Actively working on addressing test coverage.

htuch commented on source/common/config/utility.h:
```diff
@@ -329,6 +331,41 @@ class Utility {
                                     const ProtobufWkt::Struct& config,
                                     Protobuf::Message& out_proto);
+
+#if 0
```
?

htuch commented on api/envoy/admin/v2alpha/config_dump.proto:
```diff
@@ -178,3 +179,43 @@ message RoutesConfigDump {
   repeated DynamicRouteConfig dynamic_route_configs = 3 [(gogoproto.nullable) = false];
 }
+
+// Envoy's scoped RDS implementation fills this message with all currently loaded route
```
Thanks @AndresGuedez for this, my main thought looking at this is that the PR is too big to reasonably review and without diluting review bandwidth. Can we split the PR up further for review? 

I think we should have this master PR kept open to show how things tie together, but then separate PRs for:

1. Protos.
2. Scoped RDS configuration parser and handling logic. This can just be unit tested and doesn't need to be wired up yet.
3. Config subscription and wiring of scoped RDS into the rest of Envoy. The integration test can be added here.
4. Optionally, other PRs to reduce the amount of renaming noise, where reasonable.

I think we can get better review velocity if we do it this way.

AndresGuedez replied:
Ack. Working on splitting this up.

mattklein123 commented:
I think this is done and can be closed.
